Dual-Stack Integration with Kubeadm

Arvinderpal opened the pull request:
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
The PR brings necessary dual-stack related changes to --pod-network-cidr handling in kubeadm. 
See https://github.com/kubernetes/kubeadm/issues/1612 for the complete list of kubeadm changes. 

Implements:
- [x] `--pod-network-cidr` support a comma separated list of pod CIDRs. This flag is passed to the kube-controller-manager and kube-proxy. 

For building a test dual-stack cluster using kubeadm, within the kubeadm config, we must enable this feature. For example:
```
---
kind: ClusterConfiguration
featureGates:
  IPv6DualStack=true
```
For a vagrant based test cluster, please see the `dual-stack` branch of this project (https://github.com/Nordix/k8s-ipv6/tree/dual-stack) or DM on slack.

**Does this PR introduce a user-facing change?**:
```release-note
kubeadm: use the --pod-network-cidr flag to init or use the podSubnet field in the kubeadm config to pass a comma separated list of pod CIDRs. 
```

Arvinderpal commented:
/assign @yastij 
/assign @aojea

Arvinderpal commented:
/sig cluster-lifecycle

yastij commented:
/priority important-soon

aojea commented on cmd/kubeadm/app/apis/kubeadm/validation/validation.go:
```diff
@@ -381,13 +381,25 @@ func ValidateIPNetFromString(subnet string, minAddrs int64, fldPath *field.Path)
+	allErrs = append(allErrs, ValidateServiceSubnetFromString(c.ServiceSubnet, constants.MinimumAddressesInServiceSubnet, field.NewPath("serviceSubnet"))...)
 	if len(c.PodSubnet) != 0 {
-		allErrs = append(allErrs, ValidateIPNetFromString(c.PodSubnet, constants.MinimumAddressesInServiceSubnet, field.NewPath("podSubnet"))...)
+		allErrs = append(allErrs, ValidatePodSubnetsFromString(c.PodSubnet, field.NewPath("podSubnet"))...)
```
this is only going to be a comma separated list if the feature is enable. right?
` if utilfeature.DefaultFeatureGate.Enabled(kubefeatures.IPv6DualStack) ` , otherwise it has to work as always

neolit123 replied:
so on slack i said "maybe we can avoid feature gates for this", but we probably shouldn't given we are toggling an alpha feature for a number of components that can cause problems that we cannot predict:
https://github.com/kubernetes/kubernetes/blob/master/pkg/features/kube_features.go#L538

kubeadm maintains it's own way of doing feature gates (for reasons).
please, have a look at:
https://github.com/kubernetes/kubernetes/blob/master/cmd/kubeadm/app/features/features.go#L41
the unit tests should provide an example of how to use this?

kubeadm feature gates are supported in the ClusterConfiguration object (as long as they are registered), they are also supported on the CLI side but we need to double check if all `init` phases have the `--feature-gates` flag propagated.

we also need to organize the tasks better:
- first PR that we merge should add the kubeadm gate (should have a release note about the gate addition)
- second PR (this one) should modify the validation to support ",". examine if validation changes are needed based on the gate.
- third PR should add e2e tests under "kubeadm-e2e"

more comments on the implementation:
- as mentioned above validators should/might behave differently, but in my thinking they should not:
  - passing a comma separated list from the config or CLI should be tolerated by the new validators
  - a comma separated list is backwards compatible and a super-set of field that has a single value, therefore the new validators should can a superset of the of previous ones.
  - this lets us get away with supporting "," separated lists in the old configs - e.g. v1beta2
- if the kubeadm gate is enabled all components such as kubelet, control-plane, need the respected k8s feature gate to be passed to them.
- we need to maintain the kubeadm gate until the feature is GA. once it's GA we have to deprecate the feature gate.
- if we decide to deprecate the feature gate we need to follow the deprecation policy of the "state" it's currently at e.g. longer for GA.

@Arvinderpal please ping me on the PR that adds the kubeadm feature gate.
thanks!

Arvinderpal replied:
> so on slack i said "maybe we can avoid feature gates for this", but we probably shouldn't given we are toggling an alpha feature for a number of components that can cause problems that we cannot predict:

Ok. I will look at adding feature-gate in kubeadm as well. 

> we also need to organize the tasks better:
> 
> * first PR that we merge should add the kubeadm gate (should have a release note about the gate addition)
> * second PR (this one) should modify the validation to support ",". examine if validation changes are needed based on the gate.
> * third PR should add e2e tests under "kubeadm-e2e"

Ok. I have updated the tracking ticket and will use that to track current and future tasks. https://github.com/kubernetes/kubeadm/issues/1612

 
> @Arvinderpal please ping me on the PR that adds the kubeadm feature gate.

Will do! Thanks.

aojea commented on cmd/kubeadm/app/apis/kubeadm/validation/validation.go:
```diff
@@ -366,8 +366,8 @@ func ValidateHostPort(endpoint string, fldPath *field.Path) field.ErrorList {
 	return allErrs
 }
 
-// ValidateIPNetFromString validates network portion of ip address
```
maybe you can keep this function and continue using it for validation of the string

aojea commented on cmd/kubeadm/app/apis/kubeadm/validation/validation.go:
```diff
@@ -381,13 +381,25 @@ func ValidateIPNetFromString(subnet string, minAddrs int64, fldPath *field.Path)
 }
 
+// ValidatePodSubnetsFromString validates network portion of pod subnet(s)
+func ValidatePodSubnetsFromString(cidrStr string, fldPath *field.Path) field.ErrorList {
```
you should call only this function if dual stack is enabled, you also have to check that there are 2 different families , I think that something like this https://github.com/kubernetes/kubernetes/blob/54851658f3706019639c258ea0a4975e8737f2e8/test/e2e/network/dual_stack.go#L257-L260

aojea replied:
my bad, sorry, reading the previous comment I see that kubeadm doesn´t validate on feature gates 😓

aojea commented on cmd/kubeadm/app/phases/controlplane/manifests.go:
```diff
@@ -293,10 +293,10 @@ func getControllerManagerCommand(cfg *kubeadmapi.ClusterConfiguration, k8sVersio
 	// Each node will get a subspace of the address CIDR provided with --pod-network-cidr.
 	if cfg.Networking.PodSubnet != "" {
-		maskSize := calcNodeCidrSize(cfg.Networking.PodSubnet)
+		maskSize := calcNodeCidrSize(cfg.Networking.PodSubnet) // TODO(awander): Needs to be fixed once PR #73977 lands
```
I think that comments should go alone in the previous line, at least I find them easier to read

aojea commented on cmd/kubeadm/app/preflight/checks.go:
```diff
@@ -908,7 +908,11 @@ func RunInitNodeChecks(execer utilsexec.Interface, cfg *kubeadmapi.InitConfigura
-		HTTPProxyCIDRCheck{Proto: "https", CIDR: cfg.Networking.PodSubnet},
+	}
+
+	cidrs := strings.Split(cfg.Networking.PodSubnet, ",")
```
only if dualstack is enabled

yastij replied:
+1 this should be feature-gated

aojea commented on pkg/proxy/apis/config/validation/validation.go:
```diff
@@ -67,8 +67,11 @@ func Validate(config *kubeproxyconfig.KubeProxyConfiguration) field.ErrorList {
 	if config.ClusterCIDR != "" {
-		if _, _, err := net.ParseCIDR(config.ClusterCIDR); err != nil {
-			allErrs = append(allErrs, field.Invalid(newPath.Child("ClusterCIDR"), config.ClusterCIDR, "must be a valid CIDR block (e.g. 10.100.0.0/16)"))
+		cidrs := strings.Split(config.ClusterCIDR, ",")
```
only if dual stack is  enable

aojea commented on cmd/kubeadm/app/phases/controlplane/manifests.go:
```diff
@@ -293,10 +293,10 @@ func getControllerManagerCommand(cfg *kubeadmapi.ClusterConfiguration, k8sVersio
 		defaultArguments["allocate-node-cidrs"] = "true"
 		defaultArguments["cluster-cidr"] = cfg.Networking.PodSubnet
-		defaultArguments["node-cidr-mask-size"] = maskSize
+		defaultArguments["node-cidr-mask-size"] = maskSize // TODO(awander): Should be a list of maskSizes
```
I don't remember exactly, but you should check this, don't know if the first PR will have a limitation regarding this topic

Arvinderpal commented on pkg/proxy/apis/config/validation/validation.go:
```diff
@@ -67,8 +68,17 @@ func Validate(config *kubeproxyconfig.KubeProxyConfiguration) field.ErrorList {
 	if config.ClusterCIDR != "" {
-		if _, _, err := net.ParseCIDR(config.ClusterCIDR); err != nil {
-			allErrs = append(allErrs, field.Invalid(newPath.Child("ClusterCIDR"), config.ClusterCIDR, "must be a valid CIDR block (e.g. 10.100.0.0/16)"))
+		if config.FeatureGates[string(kubefeatures.IPv6DualStack)] {
```
@aojea I added the feature-gate check here since the kubeproxy config supports it.

Arvinderpal replied:
I also moved this into a separate commit since it touches kube-proxy code.

aojea replied:
@Arvinderpal @vllry seems you are changing the same file https://github.com/kubernetes/kubernetes/pull/79576 , 
I guess that changes in pkg/proxy/apis/config/validation/validation.go makes more sense in @vllry PR, what do you think?

Arvinderpal replied:
Good point. 
@vllry your PR is a lot larger and my guess is that it will likely take longer to merge into master. Would you be okay with this change being part of a kubeadm PR?
I'm planning a follow up PR for kubeadm changes related to kube-proxy. For example, we'll need to support a slice for `--bind-address` config. If you like, I can move this change to that PR instead.

neolit123 replied:
cc @rosti
patching the validation of the kube-proxy of the internal types seems required.
separate commit seems fine by me.

neolit123 replied:
the related PR seems to require rebase, so seems fine that we add it as part of this commit, for now.

Arvinderpal committed:
kubeadm --pod-network-cidr supports a comma separated list of pod
CIDRs. This is a necesary change for dual-stack.

Arvinderpal committed:
Update kubeproxy config validation to support list of comma
separated pod CIDRs. Dual-stack feature must be enabled for the
validation to be done.

yastij commented on cmd/kubeadm/app/phases/controlplane/manifests.go:
```diff
@@ -292,9 +292,11 @@ func getControllerManagerCommand(cfg *kubeadmapi.ClusterConfiguration, k8sVersio
 		maskSize := calcNodeCidrSize(cfg.Networking.PodSubnet)
 		defaultArguments["allocate-node-cidrs"] = "true"
 		defaultArguments["cluster-cidr"] = cfg.Networking.PodSubnet
+		// TODO(awander): Should be a list of maskSizes
```
can you add the reason (once we move to beta)

yastij reviewed (commented):
did a first pass

yastij commented on cmd/kubeadm/app/apis/kubeadm/validation/validation.go:
```diff
@@ -368,8 +368,8 @@ func ValidateHostPort(endpoint string, fldPath *field.Path) field.ErrorList {
-// ValidateIPNetFromString validates network portion of ip address
-func ValidateIPNetFromString(subnet string, minAddrs int64, fldPath *field.Path) field.ErrorList {
+// ValidateServiceSubnetFromString validates network portion of service subnet
+func ValidateServiceSubnetFromString(subnet string, minAddrs int64, fldPath *field.Path) field.ErrorList {
```
For now we don't have feature-gates used in kubeadm validation, I'd expect that`ValidateServiceSubnetFromString` and `ValidatePodSubnetsFromString` work as expected in both cases

cc @timothysc @neolit123

neolit123 replied:
will comment on feature gates in the bellow thread under ValidateNetworking.

yastij commented:
@Arvinderpal - can you add tests to ensure manifests generation happens correctly depending on wether we're using dual-stack or not ?

timothysc commented:
/assign @neolit123 @fabriziopandini @timothysc @kad

timothysc reviewed (changes requested):
Before I even dig too far into this, where are the tests?  

At least a set of e2e's to verify behavior even if the are [Feature:] blocked.

Arvinderpal commented:
> Before I even dig too far into this, where are the tests?
> 
> At least a set of e2e's to verify behavior even if the are [Feature:] blocked.
@timothysc Please see the e2e test that I just added.
I noticed that e2e_kubeadm actually has no tests for network configuration values. I think having such tests would be very valuable -- for example, such tests would have caught this bug: https://github.com/kubernetes/kubeadm/issues/1591

I would love to add more tests! Please let me know if the approach that I've taken here is sound. I will go ahead add additional tests related to networking configurations and dual-stack changes. :)

fabriziopandini reviewed (commented):
@Arvinderpal thanks for this PR!
Two initial comments from my side:

_Feature Flags & Validation_
Kubeadm feature flags are different from components feature flags.
I don't think we should add a new feature flag in kubeadm specific for dual stack; nevertheless, I would like to add a validation check that ensures the configuration is consistent in case that when the users opt-in into dual stack, or blocks if the users are providing inconsistent configurations. e.g. Pass if there are two subnets in all the required settings, feature flags are enabled for core components etcd.

_E2E tests_
E2E kubeadm is a test suite that validates that kubeadm makes all the expected settings in a cluster; it is ok to extend the set of tests in E2E kubeadm, but it is necessary to think how to create a periodic job that creates a dual-stack environment, spins up a cluster with the dual stack settings, and then runs E2E kubeadm (and possibly conformance)
@aojea kind can be used for this ^^

Arvinderpal commented:
@fabriziopandini 

> I don't think we should add a new feature flag in kubeadm specific for dual stack; nevertheless, I would like to add a validation check that ensures the configuration is consistent in case that when the users opt-in into dual stack, or blocks if the users are providing inconsistent configurations. e.g. Pass if there are two subnets in all the required settings, feature flags are enabled for core components etcd.

I can add validation on the networking configurations, so that they are consistent for dual-stack. However, do you think it's a good idea to add feature flag checks in kubeadm? I see two problems: (i) it may require modifying high-level funcs like ValidateInitConfiguration() since per component feature flags are spread everywhere in the configs and (ii) feature flags are temporary; eventually we'll need to remove this code. Also, I don't see any precedence in kubeadm where feature flags are being validated. 
 
> _E2E tests_
> E2E kubeadm is a test suite that validates that kubeadm makes all the expected settings in a cluster; it is ok to extend the set of tests in E2E kubeadm, but it is necessary to think how to create a periodic job that creates a dual-stack environment, spins up a cluster with the dual stack settings, and then runs E2E kubeadm (and possibly conformance)
> @aojea kind can be used for this ^^

On it! :)

neolit123 commented:
@Arvinderpal let's chat in the in #sig-cluster-lifecycle slack channel.

neolit123 commented:
xref https://github.com/kubernetes/kubeadm/issues/1612

@Arvinderpal 
a couple of related questions:
do all the kubeadm changes that we have to make target components that already support the dual-stack? is there something missing or is everything in place already - where can we see a spreadsheet or a tracking issue for this work?

Arvinderpal commented:
> do all the kubeadm changes that we have to make target components that already support the dual-stack? is there something missing or is everything in place already - where can we see a spreadsheet or a tracking issue for this work?

@neolit123 
No, the changes are trickling in gradually. I suspect this will continue for the next several months. As those PRs hit master, I will create kubeadm PRs that reflect those changes. There is no spreadsheet as of yet on the mainline dual-stack work; however, sig-network [meeting notes](https://docs.google.com/document/d/1_w77-zG_Xj0zYvEMfQZTQ-wPP4kXkpGD8smVtW_qqWM/edit) gets updated weekly with dual-stack status (e.g. see July 11th notes)

For the kubeadm changes, I will update the [tracking issue](https://github.com/kubernetes/kubeadm/issues/1612) with new tasks. Currently, these are three tasks that I see: (1) kubeadm feature-gate for dual-stack (2) podSubnet "," separated list support and (3) e2e tests. 
I would prefer not to create a spreadsheet at this point, but if you think it's useful, then I can.

neolit123 commented:
> (1) kubeadm feature-gate for dual-stack (2) podSubnet "," separated list support and (3) e2e tests.

this list of tasks SGTM.

> I would prefer not to create a spreadsheet at this point, but if you think it's useful, then I can.
> No, the changes are trickling in gradually. I suspect this will continue for the next several months. As those PRs hit master, I will create kubeadm PRs that reflect those changes. 

it doesn't have to be a spreadsheet but possibly a TODO list in the kubeadm tracking issue?
feature freeze is end of this month and if we release this as alpha in kubeadm, we need to keep track if some component settings for dual stack are not supported yet?

Arvinderpal commented:
> it doesn't have to be a spreadsheet but possibly a TODO list in the kubeadm tracking issue?
> feature freeze is end of this month and if we release this as alpha in kubeadm, we need to keep track if some component settings for dual stack are not supported yet?

@neolit123 
I added the list to the main kubeadm dual-stack tracking issue: https://github.com/kubernetes/kubeadm/issues/1612

neolit123 commented:
> I added the list to the main kubeadm dual-stack tracking issue: kubernetes/kubeadm#1612

thanks!

https://github.com/kubernetes/kubernetes/pull/80145 merged too so the gate is now present in kubeadm. we should start passing the similar k8s gate it to the components that support it.

Arvinderpal commented:
@neolit123 @yastij @fabriziopandini I have updated this PR. Please take a look. 

Also, kubeproxy validation touches kubeproxy component code. It's a simple change but should it be part of this PR? Without this change kubeadm will just error out and not create the cluster, so having it is important.

neolit123 commented on pkg/proxy/apis/config/validation/validation.go:
```diff
@@ -67,8 +68,17 @@ func Validate(config *kubeproxyconfig.KubeProxyConfiguration) field.ErrorList {
+					allErrs = append(allErrs, field.Invalid(newPath.Child("ClusterCIDR"), cidr, "must be a valid CIDR block (e.g. 10.100.0.0/16 or FD02::0:0:0/96)"))
+				}
+			}
+		} else {
```
so technically here the `if config.FeatureGates[string(kubefeatures.IPv6DualStack)] {` branch is not needed, until we add more logic related to the feature gate.

even for a single entry such as `config.ClusterCIDR = 10.100.0.0/16`, by using `cidrs := strings.Split(config.ClusterCIDR, ",")` and the `for _, cidr := range cidrs {` loop, single values will be handled too.

but i guess i'm ok leaving the separate branches for visibility of the new feature.

neolit123 commented on cmd/kubeadm/app/apis/kubeadm/validation/validation.go:
```diff
@@ -383,16 +388,54 @@ func ValidateIPNetFromString(subnet string, minAddrs int64, fldPath *field.Path)
+func ValidateIPNetFromStringForDualStack(subnet string, minAddrs int64, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+	subnets := strings.Split(subnet, ",")
+	if len(subnets) < 2 {
```
[2]
i think that instead of having a separate function `ValidateIPNetFromStringForDualStack` we can just handle dual stack checks in the following fashion:

- in `ValidateNetworking` we detect if feature gate is enabled
- we pass that as `bool` to `ValidateIPNetFromString`
- `ValidateIPNetFromString` checks if the subnet string is a list and if the feature gate is not enabled error out?
(it seems that we aren't checking the feature gate in validation.go).

this makes `ValidateIPNetFromString` handle both lists and single CIDRs.
and new tests can be added in `TestValidateIPNetFromString`.

WDYT?

Arvinderpal replied:
To detect if FG is enabled in `ValiateNetworking` we will need `ClusterConfiguration` passed as arg. That's not ideal IMO. 
I initially went with a very similar approach of passing in a `bool` to `ValidateNetworking` and down to `ValidateIPNetFromString`; I didn't like it b/c it bloated `ValiateIPNetFromString` and put dualstack logic in the top `ValidateClusterConfiguration`. The current approach seems a little cleaner. 
I'm fine to change it though. What do others think?

fabriziopandini replied:
I have some concern about this approach...
What if someone calls `ValidateNetworking` directly? it will always assume `isDualStackNetworkingConfig == false`

yastij replied:
+1 @neolit123 - I'd prefer us checking  the feature gates on `ValidateNetworking` and passing it down. also, using a global variable is not ideal.

neolit123 reviewed (commented):
thanks for the update @Arvinderpal 
i've added some comments but i think you should check [2] at the end.
perhaps we can change the logic a little, but LMK what you think.

Arvinderpal reviewed (commented):
@neolit123 Please see my comments. 
Also, separate PR for e2e test is here: https://github.com/kubernetes/kubernetes/pull/80947

aojea commented on cmd/kubeadm/app/apis/kubeadm/validation/validation.go:
```diff
@@ -383,16 +388,51 @@ func ValidateIPNetFromString(subnet string, minAddrs int64, fldPath *field.Path)
+}
+
+// isDualStackNetworking returns true if more than one pod CIDR has been specified
+func isDualStackNetworking(c *kubeadm.Networking) bool {
```
I think that the conditions is that it has more than one CIDR , and each CIDR belongs to a different IP family. Maybe we can leverage these functions in k/utils https://github.com/kubernetes/utils/blob/581e00157fb1a0435d4fac54a52d1ca1e481d60e/net/net.go#L104

fabriziopandini replied:
I'm +1 on checking more than one CIDR , and each CIDR belongs to a different IP family

Arvinderpal replied:
Done!

aojea commented on cmd/kubeadm/app/apis/kubeadm/validation/validation_test.go:
```diff
@@ -222,6 +222,36 @@ func TestValidateIPNetFromString(t *testing.T) {
+		{"valid, but IPv4 CIDR too small. At least 10 addresses needed", "10.0.0.16/29,2001:db8::/98", 10, false},
+		{"valid, but IPv6 CIDR too small. At least 10 addresses needed", "10.0.0.16/12, 2001:db8::/125", 10, false},
+		{"valid IPv4 and IPv6 CIDR", "10.0.0.16/12,2001:db8::/98", 10, true},
+		{"valid IPv6 and IPv4 CIDR", "10.0.0.16/12, 2001:db8::/98", 10, true},
```
Can we add a test case with "2 valid CIDRs but the same family"?

Arvinderpal replied:
done!

fabriziopandini commented on cmd/kubeadm/app/apis/kubeadm/validation/validation.go:
```diff
@@ -383,16 +388,51 @@ func ValidateIPNetFromString(subnet string, minAddrs int64, fldPath *field.Path)
-	allErrs = append(allErrs, ValidateIPNetFromString(c.ServiceSubnet, constants.MinimumAddressesInServiceSubnet, field.NewPath("serviceSubnet"))...)
-	if len(c.PodSubnet) != 0 {
-		allErrs = append(allErrs, ValidateIPNetFromString(c.PodSubnet, constants.MinimumAddressesInServiceSubnet, field.NewPath("podSubnet"))...)
+	if isDualStackNetworkingConfig {
```
on the same line of [2] from @neolit123 
I have some concern about this approach...
What if someone calls `ValidateNetworking` directly? it will always assume `isDualStackNetworkingConfig == false`

Arvinderpal replied:
Done!

fabriziopandini reviewed (commented):
thanks @Arvinderpal for this update, another round of comments...

Arvinderpal commented:
@neolit123 @yastij @fabriziopandini @aojea Please have another look. I have made the changes you guys suggested: 
1. We do the FG check in `ValidateNetworking` and pass it as a flag to `ValidateIPNetFromString`. 
2. `ValidateIPNetFromString` will check both single and dual stack. For dual-stack, it checks both IP families and makes use of the `k8s.io/utils/net` package. 
3. I updated unit-test `TestValidateIPNetFromString` to include dual-stack cases.

neolit123 commented on cmd/kubeadm/app/apis/kubeadm/validation/validation.go:
```diff
@@ -369,30 +371,53 @@ func ValidateHostPort(endpoint string, fldPath *field.Path) field.ErrorList {
 
 // ValidateIPNetFromString validates network portion of ip address
-func ValidateIPNetFromString(subnet string, minAddrs int64, fldPath *field.Path) field.ErrorList {
+func ValidateIPNetFromString(subnetStr string, minAddrs int64, checkDualStack bool, fldPath *field.Path) field.ErrorList {
```
can we please name the flag `isDualStackEnabled` (it's a common naming pattern).
[1]

Arvinderpal replied:
done.

neolit123 commented on cmd/kubeadm/app/apis/kubeadm/validation/validation.go:
```diff
@@ -369,30 +371,53 @@ func ValidateHostPort(endpoint string, fldPath *field.Path) field.ErrorList {
-		allErrs = append(allErrs, ValidateIPNetFromString(c.PodSubnet, constants.MinimumAddressesInServiceSubnet, field.NewPath("podSubnet"))...)
+	// check if dual-stack feature-gate is enabled
+	checkDualStack := features.Enabled(c.FeatureGates, features.IPv6DualStack)
+	// TODO(Arvinderpal): use checkDualStack flag once list of service CIDRs is supported (PR: #79386)
```
[1]

neolit123 commented on cmd/kubeadm/app/apis/kubeadm/validation/validation.go:
```diff
@@ -369,30 +371,53 @@ func ValidateHostPort(endpoint string, fldPath *field.Path) field.ErrorList {
+			isDualStack, err := utilnet.IsDualStackCIDRs(subnets)
+			if err != nil {
+				allErrs = append(allErrs, field.Invalid(fldPath, subnets, err.Error()))
+			} else if !isDualStack {
```
would this mean that if the user enables the feature gate, but the list of CIDRs they pass does not match both stacks https://github.com/kubernetes/utils/blob/c55fbcfc754a5b2ec2fbae8fb9dcac36bdba6a12/net/net.go#L80-L81 that the validation will fail.

my understanding of the feature is the following, but i could be wrong :)
- if the users passes multiple CIDRs
  - if they are from different families and if the feature gate is not enabled return error
  - if they are from the same family the feature gate is a NO-OP and we don't error out
- if the user passes a single CIDR, the feature gate is a NO-OP

please correct me, as i could be lacking some context.

Arvinderpal replied:
> * if they are from the same family the feature gate is a NO-OP and we don't error out

This is not correct. Multiple CIDRs are only supported when FG is enabled and the must contain at least one CIDR from each family.

neolit123 replied:
ok, i see.

neolit123 reviewed (commented):
@Arvinderpal please remove the WIP when you think it's OK.
leaving the LGTM to someone else.
thanks.

/approve
/ok-to-test

yastij commented on pkg/proxy/apis/config/validation/validation.go:
```diff
@@ -28,6 +28,7 @@ import (
 	"k8s.io/apimachinery/pkg/util/validation/field"
 	componentbaseconfig "k8s.io/component-base/config"
 	apivalidation "k8s.io/kubernetes/pkg/apis/core/validation"
+	kubefeatures "k8s.io/kubernetes/pkg/features"
```
this might complicate things for kubernetes/kubeadm#1600, I'd suggest dropping this import.

cc @neolit123 @rosti

neolit123 replied:
but this import is in pkg/proxy....validtion and not cmd/kubeadm.
kubeadm does not import pkg/features (last time i've checked). why do you see this as a problem?

yastij replied:
my bad, I thought I was reading on `cmd/kubeadm`

yastij replied:
@Arvinderpal - can you run `hack/update-bazel.sh` to reflect the introduced imports on the BUILD files ?

Arvinderpal replied:
Done!

neolit123 commented:
looks like there is are some failing unit tests related to the kube-proxy validation. nothing major:
https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/pull/79033/pull-kubernetes-bazel-test/1159491098037456903

Arvinderpal commented:
> Before I even dig too far into this, where are the tests?
> 
> At least a set of e2e's to verify behavior even if the are [Feature:] blocked.

@timothysc I believe I have addressed the change that you requested -- see [e2e test PR](https://github.com/kubernetes/kubernetes/pull/80947). If all looks well, then please remove the change request. :)

neolit123 commented:
@Arvinderpal this looks good on the kubeadm side, but we need an approved on the kube-proxy side:
https://github.com/kubernetes/kubernetes/blob/master/pkg/proxy/apis/config/OWNERS

looks like @thockin is the only approver in the file.
/assign @thockin 
PTAL at the `pkg/proxy/apis/config` changes, thanks!

thockin commented:
Thanks!

/approve

aojea commented:
/lgtm

ping @timothysc :)
https://github.com/kubernetes/kubernetes/pull/79033#issuecomment-520096873

neolit123 commented:
@Arvinderpal 
could you please fix the typo in release note:
`kubeamd: ` -> `kubeadm`

there is a "release-notes" tool that picks these and autogenerates a change log.
someone tried to fix it in a PR https://github.com/kubernetes/kubernetes/pull/81783

Arvinderpal commented:
@neolit123 Done. 
I assume the change is only needed in the above description and nowhere else.

neolit123 commented:
@Arvinderpal yes, AFAIK the tool operates on existing merged PRs, so having it only fixed here should be fine.
